### PR TITLE
drivers: dma buffers memory location specified in devicetree

### PR DESCRIPTION
**Pull request by awojasinski**
This PR introduces new binding for all Nordic peripherals that use EasyDMA. `dma-buffers-memory-region` is a property that allows a developer to specify the memory section where buffers passed to EasyDMA will be placed. This property requires reference to the node with compatible `zephyr,memory-region`. If `dma-buffers-memory-region` is not specified, buffers are placed in the default location.

It is the developer's responsibility to check whether the specified in the node, memory region is accessible by EasyDMA.

As an example, SPIM and UARTE shims were aligned to utilize the new feature.

### CODE REVIEW

**Comment by anangl** [comment]
cc @nika-nordic @masz-nordic

**Review by carlocaione — CHANGES_REQUESTED** [review]

File: dts/bindings/audio/nordic,nrf-pdm.yaml [hunk 1]
```diff
@@ -69,3 +69,12 @@ properties:
...
         by the driver.
+
+    dma-buffer-memory-region:
+      type: phandle
+      required: false
+      description: /
+        Memory region where EasyDMA buffer will be placed by compiler.
+        It is used by driver to place variables that are passed to EasyDMA in memory region
+        using compiler-specific feature - __attribute__((__section__(<section_name>)))
+        Phandle have to refer to node with zephyr,memory-region compatible.
```
[the remark this answers is not in the document]
Comment (carlocaione) [reply dts/bindings/audio/nordic,nrf-pdm.yaml:80]: On a second thought I think the best way is to add a more general `memory-region` property to `base.yaml` and use that instead (the property name is taken from Linux). @awojasinski-nordicsemi to be done in a separate PR though.

**Comment by JordanYates** [comment]
What is the purpose here, given that all RAM is accessible by EasyDMA?

**Comment by awojasinski-nordicsemi** [comment]
> What is the purpose here, given that all RAM is accessible by EasyDMA?

@JordanYates You may want to have a special partition or shared memory region where you want to place buffers passed to EasyDMA. This mechanism gives the programmer control over the location of buffers

**Comment by JordanYates** [comment]
> > What is the purpose here, given that all RAM is accessible by EasyDMA?
> 
> @JordanYates You may want to have a special partition or shared memory region where you want to place buffers passed to EasyDMA. This mechanism gives the programmer control over the location of buffers

I understand that's what this does, but why is that useful?
Especially given this has no impact on where the data transmitted with the UART async API or 99% of SPI data comes from.

**Comment by awojasinski-nordicsemi** [comment]
> > > What is the purpose here, given that all RAM is accessible by EasyDMA?
> > 
> > 
> > @JordanYates You may want to have a special partition or shared memory region where you want to place buffers passed to EasyDMA. This mechanism gives the programmer control over the location of buffers
> 
> I understand that's what this does, but why is that useful? Especially given this has no impact on where the data transmitted with the UART async API or 99% of SPI data comes from.

There were some cases in the past where EasyDMA didn't have access to some memory regions so it is worth having some easy-to-use mechanism that can apply a workaround. The following errata occurred in I2S but I wanted to begin with the uarte and spim shims.

https://infocenter.nordicsemi.com/index.jsp?topic=%2Ferrata_nRF52840_EngA%2FERR%2FnRF52840%2FEngineeringA%2Flatest%2Fanomaly_840_96.html&cp=4_0_1_6_1_14

**Comment by awojasinski-nordicsemi** [comment]
Closing this PR. https://github.com/zephyrproject-rtos/zephyr/pull/45142 contains all needed changes.

**Pull request closed without merging**